docs/modules/candid-guide/index.adoc

```diff
@@ -0,0 +1,1098 @@
+
+TODO
+
+[#type-principal]
+=== Type principal
+
+The Internet computer uses _principals_ as the common scheme to idnetify canisters, users and other entities.
+
+Type syntax::
+
+`principal`
+
+
+Textual syntax::
++
+....
+principal "w7x7r-cok77-xa"
+principal "zwigo-aiaaa-aaaaa-qaa3a-cai"
+principal "aaaaa-aa"
+....
+
+
+Corresponding Motoko type::
+
+`Principal`
+
+Corresponding Rust type::
+
+TODO
+
+Corresponding JavaScript values::
+
+TODO
+
+[#type-null]
+=== Type null
+
+The `null` type is the type of the value `null`, and a subtype of all the `opt t` types. It is also the idiomatic choice when using <<type-variant,variants>> to model enumerations.
+
+Type syntax::
+
+`null`
+
+Textual syntax::
+
+`null`
+
+Supertypes::
+
+All `opt t` types.
+
+Corresponding Motoko type::
+
+`Null`
+
+Corresponding Rust type::
+
+TODO
+
+Corresponding JavaScript values::
+
+TODO
+
+
+[#type-reserved]
+=== Type reserved
+
+The `reserved` type is a type with one (uninformative) value `reserved`, and is the supertype of all other types.
+
+Type syntax::
+
+`reserved`
+
+Textual syntax::
+
+`reserved`
+
+Subtypes::
+
+Any type
+
+Corresponding Motoko type::
+
+`Any`
+
+Corresponding Rust type::
+
+TODO
+
+Corresponding JavaScript values::
+
+TODO
+
+The `reserved` type can be used to remove method arguments. Consider a method with signature
+....
+service { foo : (first_name : text, middle_name : text, last_name : text) -> () }
+....
+and assume you no longer care about the `middle_name`. Although Candid will not prevent you from changing the signature to
+....
+service { foo : (first_name : text, last_name : text) -> () }
+....
+it would be disastrous: If a client talks to you using the old interface, you will silently ignore the `last_name` and take the `middle_name` as the `last_name`. Remember that method parameter names are just convention, and method arguments are identified by their position.
+
+Instead, you can use
+....
+service { foo : (first_name : text, middle_name : reserved, last_name : text) -> () }
+....
+to indicate that `foo` used to take a second argument, but you no longer care about that.
+
+
+You can also avoid this problem if you use the idiom where function with possibly changing arguments take a single record, i.e.
+....
+service { foo : (record { first_name : text; middle_name : text; last_name : text}) -> () }
+....
+Now, changing that to
+....
+service { foo : (record { first_name : text; last_name : text}) -> () }
+....
+does the right thing, and you don’t even need to keep a record of the removed argument around.
+
+
+
+[#type-empty]
+=== Type empty
+
+The `empty` type is the type without values, and is the subtype of any other type.
+
+Type syntax::
+
+`empty`
+
+Textual syntax::
+
+None, as this type has no values
+
+Supertypes::
+
+Any type
+
+Corresponding Motoko type::
+
+`None`
+
+Corresponding Rust type::
+
+TODO
+
+Corresponding JavaScript values::
+
+TODO
+
+Practical uses of the `empty` type are relatively rare.
```

Comment:
One use case is for expressing an (output) record field that is no longer needed, by upgrading
```
record {foo : nat}  -->  record {foo : empty}
```

Reply:
That doesn’t work, right? Because now the whole `record` is empty. This way you can mark methods as “I will never return any more”, but then you just use `empty` instead of `record { foo : empty }`.

Reply:
You could use it to declare that you stopped producing a certain _variant_ tag, but then you could just remove that variant…

Reply:
Ah yes, sorry, I was thinking about the case where you have an option, should have been `...opt nat --> ...opt empty`. But admittedly that is a narrow use case.